src/Cmts.ml

```diff
@@ -402,29 +393,34 @@ let init_use_file =
              List.map items ~f:(fun {Parsetree.pstr_loc} -> pstr_loc)
          | Ptop_dir _ -> [] ))
 
-(** Relocate comments, for Ast transformations such as sugaring. *)
-let relocate t ~src ~before ~after =
-  let update_multi tbl src dst ~f =
-    Option.iter (Hashtbl.find_and_remove tbl src) ~f:(fun src_data ->
-        Hashtbl.update tbl dst ~f:(fun dst_data ->
-            Option.fold dst_data ~init:src_data ~f:(fun src_data dst_data ->
-                f src_data dst_data ) ) )
-  in
-  if Conf.debug then
-    Format.eprintf "relocate %a to %a and %a@\n%!" Location.fmt src
-      Location.fmt before Location.fmt after ;
-  update_multi t.cmts_before src before ~f:(fun src_cmts dst_cmts ->
-      List.append src_cmts dst_cmts ) ;
-  update_multi t.cmts_after src after ~f:(fun src_cmts dst_cmts ->
-      List.append dst_cmts src_cmts )
-
 let remove = ref true
 
-let preserve f x =
+let preserve fmt_x x =
+  let buf = Buffer.create 128 in
```

Comment:
@gpetiot the previous version was buggy in that it always used the same buffer, which for nested expressions was just wrong